engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/FormResourceTransformer.java

```diff
@@ -0,0 +1,172 @@
+import io.camunda.zeebe.protocol.record.intent.FormIntent;
+import io.camunda.zeebe.stream.api.state.KeyGenerator;
+import io.camunda.zeebe.util.Either;
+import java.io.IOException;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import org.agrona.DirectBuffer;
+
+public final class FormResourceTransformer implements DeploymentResourceTransformer {
+
+  private static final int INITIAL_VERSION = 1;
+
+  private static final Either<Failure, Object> NO_DUPLICATES = Either.right(null);
+
+  private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
+
+  private final KeyGenerator keyGenerator;
+  private final StateWriter stateWriter;
+  private final Function<DeploymentResource, DirectBuffer> checksumGenerator;
+  private final FormState formState;
+
+  public FormResourceTransformer(
+      final KeyGenerator keyGenerator,
+      final StateWriter stateWriter,
+      final Function<DeploymentResource, DirectBuffer> checksumGenerator,
+      final FormState formState) {
+    this.keyGenerator = keyGenerator;
+    this.stateWriter = stateWriter;
+    this.checksumGenerator = checksumGenerator;
+    this.formState = formState;
+  }
+
+  @Override
+  public Either<Failure, Void> transformResource(
+      final DeploymentResource resource, final DeploymentRecord deployment) {
+
+    return parseFormId(resource)
+        .flatMap(
+            formId ->
+                checkForDuplicateFormId(formId, resource, deployment)
+                    .map(
+                        noDuplicates -> {
+                          final FormMetadataRecord formRecord = deployment.formMetadata().add();
+                          appendMetadataToFormRecord(formRecord, formId, resource);
+                          writeFormRecord(formRecord, resource);
+
+                          return null;
+                        }));
+  }
+
+  private Either<Failure, String> parseFormId(final DeploymentResource resource) {
+    try {
+      final String formId = JSON_MAPPER.readValue(resource.getResource(), FormIdPOJO.class).getId();
+
+      return validateFormId(formId);
+    } catch (final JsonProcessingException e) {
+      final var failureMessage =
+          String.format(
+              "Failed to parse formId from form JSON. '%s': %s",
+              resource.getResourceName(), e.getCause().getMessage());
+      return Either.left(new Failure(failureMessage));
+    } catch (final IOException e) {
+      final var failureMessage =
+          String.format("'%s': %s", resource.getResourceName(), e.getCause().getMessage());
+      return Either.left(new Failure(failureMessage));
+    }
+  }
+
+  private Either<Failure, ?> checkForDuplicateFormId(
+      final String formId, final DeploymentResource resource, final DeploymentRecord record) {
+    return record.getFormMetadata().stream()
+        .filter(metadata -> metadata.getFormId().equals(formId))
+        .findFirst()
+        .map(
+            previousResource -> {
```

Comment:
❌ This is not the previous resource, I suggest to change the name as follow:

```suggestion
            duplicatedForm -> {
```

Reply:
Good one!